_posts/2018-08-30-pair-programming.markdown

```diff
@@ -0,0 +1,184 @@
+---
+layout: epic
+title: "The Hows and Whys of Pair Programming"
+date: 2018-08-30
+author: [yuki, orta]
+categories: [pairing, culture]
+---
+
+Why pair program? As new engineers join Artsy we've been experimenting with different programming cultures - Yuki
+came from Pivotal Labs where they have a strong pair programming culture and introduced it at Artsy - it's been
+about a year and a half and we're all really loving the changes he's introduced.
+
+I asked Yuki if he'd pair program with me on a blog, post on Pair Programming, and this is it. So, we're going to
+dive into what pair programming is, why you should do it, what are good mental models to think about, the techniques
+you can use to make it work, what hardware you might need and how Yuki persuaded so many of us to start doing it
+more often.
+
+<!-- more -->
+
+## Why pair program?
+
+The idea of pair programming came from a very popular book called [Extreme Programming Explained: Embrace
+Change][extreme-explained]. Code review has been a good practice in software development. Many organizations have
+adopted it, but XP (Extreme Programming) literally takes this **more extremely** - what if the engineer sitting next
+to you is reviewing your code as soon as you write it? This requires two engineers to simultaneously pay attention
+to the code that's being written, discuss implementation details, and sometimes socialize. That's how pair
+programming was born.
+
+Pair programming seems as easy as it sounds, and to some extent that's true. However, with just a few key points
+you'll be able to make your pairing session much more valuable.
+
+To think about when pairing is most valuable, let's think about code review first. Modern software organizations
+have implemented a process where engineers send pull requests to get reviews on GitHub (or whatever tool your
+organization uses for reviewing code). This is a very powerful process, but most of us have probably experienced a
+situation where a pull request with a very simple change created a very long discussion. Sometimes
```

Comment:
Any (in)famous examples we could link to?

> created a very long discussion

Reply:
One example would be [this PR that added our `<Collapse>` component to Reaction](https://github.com/artsy/reaction/pull/1114). The collapse animation is very common and the whole change is simple enough. I'd expected a light and quick review, but we ended up having a very long discussion with 25 comments from 9 people including a designer. The other example on an actual open source project is [this PR on Rails](https://github.com/rails/rails/pull/23012). That's literally 2 lines of diff, but there are 20 comments in there, a few of which are very long.

@ashfurrow what do you think is a good example?

Reply:
I remember that Collapse PR! Let's use that one, since it's Artsy's and will help normalize the behaviour of teams reflecting on themselves and their processes.